### PR TITLE
Identity Deposits Relay to Parachain Migration

### PR DESCRIPTION
**Pull request by joepetrowski**
The goal of this PR is to migrate Identity deposits from the Relay Chain to a system parachain.

The problem I want to solve is that `IdentityOf` and `SubsOf` both store an amount that's held in reserve as a storage deposit. When migrating to a parachain, we can take a snapshot of the actual `IdentityInfo` and sub-account mappings, but should migrate (off chain) the `deposit`s to zero, since the chain (and by extension, accounts) won't have any funds at genesis.

The good news is that we expect parachain deposits to be significantly lower (possibly 100x) on the parachain. That is, a deposit of 21 DOT on the Relay Chain would need 0.21 DOT on a parachain. This PR proposes to migrate the deposits in the following way:

1. Introduces a new pallet with two extrinsics: 
    - `reap_identity`: Has a configurable `ReapOrigin`, which would be set to `EnsureSigned` on the Relay Chain (i.e. callable by anyone) and `EnsureRoot` on the parachain (we don't want identities reaped from there).
    - `poke_deposit`: Checks what deposit the pallet holds (at genesis, zero) and attempts to update the amount based on the calculated deposit for storage data.
2. `reap_identity` clears all storage data for a `target` account and unreserves their deposit.
3. A `ReapIdentityHandler` teleports the necessary DOT to the parachain and calls `poke_deposit`. Since the parachain deposit is much lower, and was just unreserved, we know we have enough.

One awkwardness I ran into was that the XCMv3 instruction set does not provide a way for the system to teleport assets without a fee being deducted on reception. Users shouldn't have to pay a fee for the system to migrate their info to a more efficient location. So I wrote my own program and did the `InitiateTeleport` accounting on my own to send a program with `UnpaidExecution`. Have discussed an `InitiateUnpaidTeleport` instruction with @franciscoaguirre . Obviously any chain executing this would have to pass a `Barrier` for free execution.

TODO:

- [x] Confirm People Chain ParaId
- [x] Confirm People Chain deposit rates (determined in https://github.com/paritytech/polkadot-sdk/pull/2281)
- [x] Add pallet to Westend

### CODE REVIEW

**Comment by burdges** [comment]
Identity could work with only state proofs, so without messages, but messages should be simpler initially.

File: substrate/frame/identity/src/lib.rs
Comment (georgepisaltu) [inline substrate/frame/identity/src/lib.rs:1006]: Why do we need to do this check? Is there any scenario where this function would be called with `num_subs` not being equal to `T::MaxSubAccounts`? We can make it account for that with weight refunds if necessary because IIUC in the current impl the caller will pay proportional to their `num_subs` of choice regardless of how many subs there actually are (and I doubt callers want to check each identity to see the exact numbers).

File: substrate/frame/identity/src/lib.rs
Comment (georgepisaltu) [inline substrate/frame/identity/src/lib.rs:1060]:
```suggestion
			Self::rejig_deposit(&target, current_subs_deposit, new_subs_deposit)?;
```

File: substrate/frame/identity/src/lib.rs
Comment (georgepisaltu) [inline substrate/frame/identity/src/lib.rs:1058]: Let's use `SubsOf::<T>::try_mutate` instead of `take` and `insert`. The code below just changes the value in the map, but the entry stays.

File: polkadot/runtime/rococo/src/impls.rs
Comment (georgepisaltu) [inline polkadot/runtime/rococo/src/impls.rs:42]: I'd like to have these moved outside this function to constants and...

File: polkadot/runtime/rococo/src/impls.rs
Comment (georgepisaltu) [inline polkadot/runtime/rococo/src/impls.rs:50]: ... and have this computation in a separate function which uses those constants
```
impl ToParachainIdentityReaper {
fn total_amount(additional_fields: usize, subs: usize) -> Balance
}
```
to segregate that from the actual XCM stuff here.

**Review by georgepisaltu — COMMENTED** [review]
First pass, I haven't yet looked into the tests and benches but the approach looks solid to me.

File: polkadot/runtime/rococo/src/impls.rs
Comment (joepetrowski) [reply polkadot/runtime/rococo/src/impls.rs:42]: Yes that's the plan. Once there is a people/identity chain runtime, we can just read the constants right from that.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3923274

File: substrate/frame/identity/src/lib.rs
Comment (georgepisaltu) [reply substrate/frame/identity/src/lib.rs:1032]: I favor option 1 over option 2 and the reasons are:
- I don't see a scenario where I'd like to reap an identity but only if the number of subs is less than some arbitrary bound I set. In other words, it looks from the outside like `num_subs` is a protection for the user against excessive weight if they don't want to look into the identity they want to reap, which is not a problem if we do weight refunds.
- It would make sense to have the callers look into the struct off-chain if it was computationally intensive, but it's not the case here since we already have access to the number of subs just by accessing the list. We can do this for the user cheaply, it integrates into the system nicely with the weight refund, and it's a simpler interface IMO.

I'm not going to block on this. Just giving my 2c.

File: substrate/frame/identity/src/lib.rs
Comment (georgepisaltu) [inline substrate/frame/identity/src/lib.rs:1107]: In order to be as explicit as possible with this, we can do `try_mutate` instead of `put` and return an error in case the flag was already `true`. Not a requirement, but more defensive. Same for the unlocking extrinsic.

Comment (joepetrowski) [reply substrate/frame/identity/src/lib.rs:1109]: Returning an error and putting the same value are both no-ops though. What would it defend against?

Comment (georgepisaltu) [reply substrate/frame/identity/src/lib.rs:1109]: While the effects in storage would be the same, if you run `lock_pallet` with the flag already set, it's more obvious/alarming to the caller if they get an error that something's not right, because they'd call that only if they thought the pallet wasn't already locked. Functionally it makes no difference, the fees would be charged anyway. I'm fine with whatever you decide here.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4062158

File: substrate/frame/identity/src/lib.rs
Comment (bkchr) [inline substrate/frame/identity/src/lib.rs:1071]: To lock the pallet, we should just remove the possibility to call this pallet from the runtime. Not requiring any special logic in this pallet here. This can be done by just removing `Call` in the `Identity` line in `construct_runtime`.

File: substrate/frame/identity/src/lib.rs
Comment (joepetrowski) [reply substrate/frame/identity/src/lib.rs:979]: Because we have to "migrate" the deposits to a new runtime, and this pallet holds funds in reserve. To migrate that directly, it would mean minting funds in the genesis state of the new parachain, and burning funds in a Relay Chain runtime upgrade. IMO that is not very transparent; people should be able to see a transaction that moved their funds from one place to another.

That is the purpose of the `ReapIdentityHandler`: In the case of migrating to a parachain, it will teleport the funds needed for the deposit there. This will make the migration of funds transparent. It will also allow parachain deposit calculation to happen in the handler to ensure that the correct amount is teleported there.

File: substrate/frame/identity/src/lib.rs
Comment (joepetrowski) [reply substrate/frame/identity/src/lib.rs:1071]: That was my initial approach but I think that is wrong for two reasons:

1. We still want to be able to call `reap_identity`, and
2. It will result in a lot of downtime because the unlocking action will need to be done via runtime upgrade. I was going to propose that the Fellowship be the `LockerOrigin` such that it could unlock the parachain once all deposits have been migrated. We can then change it to Root.

Comment (bkchr) [reply substrate/frame/identity/src/lib.rs:1071]: > 2\. such that it could unlock the parachain once all deposits have been migrated

But this will require on the users actually doing this? Sounds like it takes quite some time?

Comment (joepetrowski) [reply substrate/frame/identity/src/lib.rs:1071]: No, once the parachain launches where we want the deposits to migrate to, the `ReapOrigin` would be set to `EnsureSigned` so that anyone could call this and migrate every account. See the docs on the [Rococo Config](https://github.com/paritytech/polkadot-sdk/pull/1814/files#diff-da220c6c064cc047fef5d3b4a594ae30b8bbf48a29b740742bc9858285c7ba57R623).

Since the `IdentityInfo` and other state will be in the parachain's genesis (with the pallet locked at genesis), we can very quickly migrate all the deposits to the parachain and unlock the pallet there. Then we just remove the pallet from the Relay Chain entirely in the next runtime upgrade.

**Review by muharem — COMMENTED** [review]
Given these constraints, the current approach seems sound to me. However, I'd like to revisit them and offer an alternative solution for consideration:

- user-friendliness: users shouldn't have to take any action for this migration.
- on-chain accountability for deposit transfers from the relay chain to the parachain.
- minimal downtime for write operations in the identity pallet.

to honor the first constraint, we must handle data migration seamlessly and introduce calls like `reap_identity` with a target account, enabling actions on behalf of the user.

to meet the second constraint, deposits must be migrated via XCM. simple migration without this is not an option.

for the third constraint, we should have the ability to lock and unlock the calls for the identity pallet on the new people chain through a specific extrinsic call. it's important to note that this constraint applies only to the people chain. on the relay chain, we can deactivate all pallet calls with an upgrade dedicated for the migration and completely remove the pallet with a subsequent upgrade.

consider this: what if we think it's acceptable to keep all identity pallet's calls deactivated on the people chain until the next upgrade, or if we can expedite that upgrade?

if so, we could introduce an additional custom pallet tailored for migration (outside of frame). this would allow the identity pallet to remain largely untouched, avoiding any breaking changes.

within the identity pallet, we'd need to unveil two public functions: one to clear identity and refund the deposit and another to calculate and take the deposit. This custom pallet would then implement `reap_identity` and `poke_deposit` using those functions.

@joepetrowski @bkchr

**Review by muharem — APPROVED** [review]
Looks good!

**Review by georgepisaltu — APPROVED** [review]
LGTM

File: substrate/frame/identity/src/lib.rs
Comment (NachoPal) [inline substrate/frame/identity/src/lib.rs:977]: Already checked inside the `try_mutate`

**Review by NachoPal — APPROVED** [review]
LGTM, just some nits

File: polkadot/runtime/rococo/src/impls.rs
Comment (joepetrowski) [reply polkadot/runtime/rococo/src/impls.rs:134]: It was a "safe guess", but I will update before merging.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4337801

File: polkadot/runtime/rococo/src/impls.rs
Comment (bkchr) [inline polkadot/runtime/rococo/src/impls.rs:98]: Here you also always only call `roc.inner().first().expect()`

**Review by bkchr — APPROVED** [review]
LGTM

File: polkadot/runtime/westend/src/impls.rs
Comment (joepetrowski) [reply polkadot/runtime/westend/src/impls.rs:98]: Yeah you are right. Before I was doing another approach that took `MultiAssets`, but we don't need it.

File: polkadot/runtime/common/src/identity_migrator.rs
Comment (xlc) [inline polkadot/runtime/common/src/identity_migrator.rs:1]: can we stop adding more code to polkadot-runtime-common https://github.com/paritytech/polkadot-sdk/issues/2186

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadot-release-analysis-v1-4-0/5152/1

**Pull request merged**
